Add fish shell support

JBlaschke opened the pull request:
As requested in #7311, this adds a `setup-env.fish` enabling the same level of user interactivity in fish as in bash. 

Since `setup-env.sh` uses quite a bit of shell magic specific to bash-like syntax, I've had to completely rewrite it for fish. The design is very much based in `setup-env.sh`. I've tested this on a few standard use cases (bootstraping, inspecting packages, installing, loading and unloading with `environment-modules`). I do welcome feedback from more rigorous tests.

Note that in order to get `environment-modules` to cooperate with fish, I had to upgrade it to >v4.

Closes #7311

JBlaschke commented:
FYI: I'm keeping my Fork up-to-date as I am using spack from within fish

michaelkuhn commented on share/spack/setup-env.sh:
```diff
@@ -245,8 +245,8 @@ if [ "${need_module}" = "yes" ]; then
-        module() { eval `${MODULE_PREFIX}/Modules/bin/modulecmd ${SPACK_SHELL} $*`; }
+        _spack_pathadd PATH "${MODULE_PREFIX}/bin"
+        module() { eval `${MODULE_PREFIX}/bin/modulecmd ${SPACK_SHELL} $*`; }
     fi;
```
It might make sense to keep this backwards-compatible, that is, test whether `Modules/bin` exists and use it if it does.

JBlaschke replied:
`Modules/bin` no longer exists with modules version 4. A solution would be to create a symlink.

michaelkuhn replied:
It should also be possible to first check whether bin exists and only if it does not, fall back to Modules/bin, right?

michaelkuhn commented on var/spack/repos/builtin/packages/environment-modules/package.py:
```diff
@@ -27,13 +27,13 @@
 
-    version('3.2.10', '8b097fdcb90c514d7540bb55a3cb90fb')
+    version('4.1.4', sha256='965b6056ea6b72b87d9352d4c1db1d7a7f9f358b9408df2689d823b932d6aa7f')
 
```
We generally like to keep old versions in the package (see also above regarding backwards compatibility).

JBlaschke replied:
Modules version 3 is not compatible with fish, that's why I updated it.

JBlaschke replied:
By the way: when you say you want to keep old versions, do you mean "as well" (i.e. spack should provide a choice of versions)? Or are you generally opposed to upgrading any packages

healther replied:
It is always an "as well", spack will default to the newest version, but it should retain the information to build old versions, as one of the main points of spack is reproducibility

JBlaschke replied:
I like that! Is there a way for me to add multiple versions of modules (i.e. keep the old one for compatibility *and* add the new one)? Sorry for the stupid question, I'm still relatively new to spack.

healther replied:
Each installed spec (i.e. each configuration of versions and variants of a package) will also produce a module. There is quite a lot of infrastructure to modify these, but you should ask @alalazo he knows this part inside out. Or maybe talk to us on slack if you need/want help

michaelkuhn commented on var/spack/repos/builtin/packages/modules/package.py:
```diff
@@ -0,0 +1,86 @@
+
+        configure(*config_args)
+        make()
+        make('install')
```
This seems to be a duplicate of the environment-modules or am I missing something?

JBlaschke replied:
I'm sorry, this kinda slipped in unintentionally. I had this package in there to experiment with different modules versions in my machine.

If we downgrade the main one back to 3.2.10, we could use this one for fish support. This will make bootstrapping more complex though.

JBlaschke commented:
I'm sorry, this kinda slipped in unintentionally. I had this package in there to experiment with different modules versions in my machine.

If we downgrade the main one back to 3.2.10, we could use this one for fish support. This will make bootstrapping more complex though.

pat-s commented:
Has there been recent activity in here? I would love to see fish support.

JBlaschke commented:
> Has there been recent activity in here? I would love to see fish support.

I'm sorry, I haven't had time these last few months to address the comments raised  by @michaelkuhn  and @healther 

I'll get back on it soon.

xdelaruelle commented:
@JBlaschke now environment-modules recipe has been updated (7eec0386) to support newer version (4+), it may simplify your work here

pat-s commented:
@JBlaschke Do you think you continue working on this? Would be great to have fish support!

JBlaschke commented:
@pat-s @xdelaruelle Hey guys, I have updated my `setup-env.fish` to include spack's updated modules recipe. It should work the same way as the bash version now. Sorry that this took so long, no excuses!

JBlaschke commented:
@heather @michaelkuhn I think my changes now have addressed your comments on the previous version. Let me know what you think.

I have two questions regarding my changes:
1) I noticed that the bash version dropped dotkit support. Should I do the same in fish? (I left some TODO comments to this effect)
2) I only did some rudimentary tests (installing, and loading modules). Are there any CI tests for `setup-env.sh` that I can adapt for `setup-env.fish`?

tgamblin commented:
FYI @gonsie 😄

pat-s commented:
Thanks a lot @JBlaschke! First few commands worked. Finally fish supp 🎉

adamjstewart commented:
> I noticed that the bash version dropped dotkit support. Should I do the same in fish? (I left some TODO comments to this effect)

Yes, you can drop dotkit support now.

adamjstewart commented:
It looks like a few unrelated commits got sucked in when you merged in master.

JBlaschke commented:
Oh crap! I didn't see that. Thanks...

JBlaschke commented:
@pat-s @tgamblin This is just a heads-up guys: I'm about done with spack's fish support.

From my end, I just want to clean up some things a little. The codecov tests that are failing are due to the fish shell support that I added to `lib/spack/spack/environment.py`. @tgamblin I will need some guidance about setting up condecov tests for those bits.

@pat-s On my system, everything's working fine. Let me know if something's broken

JBlaschke commented:
@pat-s if you have time, also have a look at `share/spack/qa/setup-env-test.fish`, and see if it works on your end

pat-s commented:
Thanks, I checked out your branch and will report any glitches I might experience.

pat-s commented:
Possibly unrelated to your implementation:

With bash I have autocomplete with `lmod` when using

```sh
module load <module>
```

with fish, I need to 

1. Have spacks `environment-modules` installed
2. fish then uses `modules` instead of `lmod` via the `modules` command
3. no autocomplete for loading modules

Actually I would like to use `lmod` with autocomplete also for fish (prefer it over `spack load`). It might be though that this behavior is not within the scope of this PR here.

---

Also after installation of a spack library, I need to refresh the terminal session before I can do `spack load <new library>`. With bash that worked without a session refresh.

JBlaschke commented:
@pat-s thanks! It looks like I was wrapping `environment-modules` regardless of `$need_module`. That was a mistake, which the latest version should have corrected. You should be able to use `lmod` without interference from spack now.

The objective of `setup-env.fish` is to emulate `setup-env.sh`'s behavior as closely as possible.

I was not able to reproduce needing to refresh the terminal session problem. Can you give me more details? If you can give me a test case I will investigate what might be the problem here.

As an example (where I don't need to do anything more to the terminal) is to install and use spack's `tar`
1. getting a clean spack
2. running `spack bootstrap` since I'm using spack's `environment-modules`... you should be able to skip this step
3. `spack find` shows:
```
==> 3 installed packages
-- darwin-mojave-x86_64 / clang@10.0.0-apple --------------------
environment-modules@4.4.0  tcl@8.6.8  zlib@1.2.11
```
4. `spack install tar`,  followed by `spack find` then shows
```
==> 5 installed packages
-- darwin-mojave-x86_64 / clang@10.0.0-apple --------------------
environment-modules@4.4.0  libiconv@1.16  tar@1.32  tcl@8.6.8  zlib@1.2.11
```
5. I can then use spack's `tar`: `spack load tar` followed by `which tar` shows:
```
/Users/johannesblaschke/Developer/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.0-apple/tar-1.32-hmvw2pv5bnmspkmvvivoncqozlwp2vgp/bin/tar
```

If step 2 above is the problem, let me know and I can try and test with `lmod`

adamjstewart commented:
@pat-s I think there's two different things here:

1. Programmable tab completion for Environment Modules/Lmod
2. Programmable tab completion for Spack

The former is definitely outside the scope of this PR. I'm sure it already exists, you just have to install it on your system.

The latter is within the scope of this PR, although I think it could wait for a follow-up PR so we can merge this quicker. @JBlaschke if you want to take a stab at this, see the `share/spack/spack-completion.bash` script I wrote. I don't know enough about Fish to tell how much work it will be to port that from Bash, but most of the code is pretty simple. Let me know if you have any questions.

pat-s commented:
With `environment-modules` NOT installed, I get

```sh
spack load lmod

fish: Unknown command '/opt/spack/opt/spack/linux-centos7-x86_64/gcc-9.2.0/environment-modules-4.4.0-ip43bkldn6wb3g22qwhdab5wwz24fstu/bin/modulecmd'
- (line 1):
begin; /opt/spack/opt/spack/linux-centos7-x86_64/gcc-9.2.0/environment-modules-4.4.0-ip43bkldn6wb3g22qwhdab5wwz24fstu/bin/modulecmd fish load lmod-8.1.5-gcc-9.2.0-iqr34oa
       ^
from sourcing file -
        called on line 61 of file /opt/spack/opt/spack/linux-centos7-x86_64/gcc-9.2.0/fish-3.0.0-gdyab6rniiipvhdqd6lo45ixszdevd6k/share/fish/functions/eval.fish

in function “eval”
        called on line 1517 of file /opt/spack/share/spack/setup-env.fish

in function “module”
        called on line 373 of file /opt/spack/share/spack/setup-env.fish
        with parameter list “load lmod-8.1.5-gcc-9.2.0-iqr34oa ”

in command substitution
        called on line 946 of file /opt/spack/share/spack/setup-env.fish

in function “spack”
        called on standard input
        with parameter list “load lmod”
```

also when calling `module` standalone I get an error due to not having `environment-modules` installed.

With `bash`, `lmod` is taking over for me - I am wondering why.
Somehow I must have `lmod` in my bash `$PATH` somehow but not available in `fish`.

Found it: On my cluster we are running also OpenHPC libs and with these we get `/etc/profile.d/lmod.sh`, which brings `lmod` support for bash.

With `fish` however, even if I post-load `lmod`, `environment-module` is still first in PATH. But this seems to be off-topic.

So it seems I just found out that it actually is not the default case to have `lmod` doing the job when calling `module` - instead `environment-modules` is what `spack` has as the default when calling `module`. 🙄 

Also since `fish` is not the default shell but just launched by `bash`, I get all the loaded modules specificed in `.bashrc` inherited into the `fish` session. 

---

So in summary, there seem to be some glitches (which are out of scope of this PR) related to using external tools for loading env modules. Thanks a ton @JBlaschke for all your work here :)

pat-s commented:
@JBlaschke Thanks for taking a stab at it. Even after doing `lmod-8.1.5-gcc-9.2.0-iqr34oa` / `spack load lmod`, `environment-modules` is still preferred over `lmod` when executing `module`.

However, I am not sure if this is really an issue if this implementation - just go on here, this can maybe also solved in a new PR.

JBlaschke commented:
@adamjstewart I agree that the completions should be a separate PR. I had a look at your bash completions and I think it will take me a couple of evenings at least to figure out how to adapt this to fish (this would be the first time that I set up completions). I'll create a new PR once I've got completions working (eta 3-ish weeks).

@adamjstewart @healther and @michaelkuhn I think this PR is ready to be merged. Let me know if you need anything more. I think the codecov/patch test fails because of my additions to `lib/spack/spack/environment.py`. I don't know enough about codecov to fix to failure though. Any help would be appreciated.

@pat-s I'll put an issue to test `spack`+`lmod` into my kanban. I'll have time to revisit this down the road. My system at home has fish as a default shell, so I don't have all the PATH problems. I did run into similar issues on NERSC, what my local fish install is launched from bash. The way I solved it was by installing [jorgebucaran/fish-bax](https://github.com/jorgebucaran/fish-bax), together with a wrapper function that passes the module command to bash:

```fish
# in config.fish
if test "$NERSC_HOST" = "cori"  # or whatever you use to check for the host
    function module
        bax "module $argv"
    end
end
```
this way you'll "inherit" the module command from bash... I hope this helps

adamjstewart commented:
> @adamjstewart I think that the completions should be a separate PR. I had a look at your bash completions and I think it will take me a couple of evenings at least to figure out how to adapt this to fish (this would be the first time that I set up completions). I'll create a new PR once I've got completions working (eta 3-ish weeks).

Sounds good to me! What would be really awesome is if we could have a common file containing all of the `_spack_{subcommands}` functions that works in bash/zsh/fish, and then have bash/zsh/fish-specific files that define `_{shell}_completion_spack`. That would mean users don't have to update 2-3 files every time a new flag is added to a command. But I don't think fish has `compgen`, so I'm not sure how that would work. If you can figure out a way to auto-generate the tab completion script, that would be even better, but I never managed to get that working.

pat-s commented:
> this way you'll "inherit" the module command from bash... I hope this helps

perfect! This is nothing but awesome 🎉

tgamblin commented:
FYI for folks on this thread: @adamjstewart is going to work on automatic completion generation based on the prototype in #14393, so it should be possible to simplify things quite a bit.

gonsie commented on share/spack/setup-env.fish:
```diff
@@ -0,0 +1,826 @@
+# also puts spack in your path. The script also checks that at least module
+# support exists, and provides suggestions if it doesn't. Source it like this:
+#
+#    . /path/to/spack/share/spack/setup-env.fish
```
```suggestion
#    source /path/to/spack/share/spack/setup-env.fish
```

The `.` alias is [deprecated in fish](https://fishshell.com/docs/current/commands.html#source).

Thank you so much for doing this!!

JBlaschke commented:
> After trying it out with `fish` as the default shell, I get
> 
> ```
> module                                                                                                                                                                                                       
> bash: line 1: module: command not found
> ```
> 
> doing `spack load lmod` changes nothing.
> 
> I'm on the latest revision of this PR and sourcing `. /opt/spack/share/spack/setup-env.fish` in `config.fish`. I'd expect that at least `environment-modules` should be loaded at this point?

@pat-s I haven't had an opportunity to test this with `lmod`.  But two things come to mind:

1) why is bash complaining when you've trying out `fish` as your default shell? If you've still got my fix using `bax` (defined [here](https://github.com/spack/spack/pull/9279#issuecomment-570998593)) then this would explain it. This fix only works if you've got fish running on top of a fully-configured bash.

2) The `share/spack/setup-env.fish` only defines a `module` function, if `$need_module` is set to "yes". One way I can see this breaking is if you've got a `module` or a `use` function defined (for example if pt 1 above is the case).

To debug this, can you add and `echo $need_module` [here](https://github.com/JBlaschke/spack/blob/051f4f62c290e90701091956b7bc8afa603e1540/share/spack/setup-env.fish#L786).

JBlaschke commented:
@tgamblin I see that @adamjstewart has made loads of progress on #14393. Once that's merged, I'll add a PR to include fish support (this should only require updating the `spack/lib/spack/spack/cmd/commands.py` file with the relevant `--format=fish` flag)

tgamblin commented:
@JBlaschke: awesome!  it's in.

tgamblin commented:
@JBlaschke: see also #14607

pat-s commented:
@JBlaschke 

> This fix only works if you've got fish running on top of a fully-configured bash.

Thanks, that was it.

adamjstewart commented on share/spack/qa/setup-env-test.fish:
```diff
@@ -0,0 +1,392 @@
+#
+
+
+function allocate_testing_global -d "allocate global variables used for testing"
```
In #14393, I split the first half of this file (the testing framework) into a separate file so that it could be sourced in both `setup-env-test.sh` and `completion-test.sh`. You can either do the same in this PR, or a follow-up PR when you add fish tab completion support.

JBlaschke replied:
Thanks for the suggestion I think I'll do it in the follow-up PR. This way we can resolve this PR.

JBlaschke replied:
Adding an issue in my fork so I don't forget

adamjstewart commented on share/spack/qa/setup-env-test.fish:
```diff
@@ -0,0 +1,392 @@
+true # ingnore failing `set -e`
+
+# Source setup-env.sh before tests
+source share/spack/setup-env.fish
```
This doesn't work unless the script is run from `SPACK_ROOT`. In #14393 I changed the test file to calculate which directory it lives in and find the right `setup-env` file.

JBlaschke replied:
Yes, you're right. In the interest in getting this PR merged, I would vote to do this change in the follow-up PR (see previous comment above), when I add fish tab completion.

JBlaschke replied:
Also adding this to my fork's issue list

adamjstewart commented on share/spack/qa/setup-env-test.fish:
```diff
@@ -0,0 +1,392 @@
+
+title "Testing setup-env.fish with $_sp_shell"
+
+# spack command is now avaialble
```
```suggestion
# spack command is now available
```

adamjstewart commented on share/spack/qa/setup-env-test.fish:
```diff
@@ -0,0 +1,392 @@
+set b_install (spack location -i b)
+set b_module (spack -m module tcl find b)
+
+# create a test environment for tesitng environment commands
```
```suggestion
# create a test environment for testing environment commands
```

adamjstewart commented on share/spack/qa/setup-env-test.fish:
```diff
@@ -0,0 +1,392 @@
+
+# Make sure no environment is active
+set -e SPACK_ENV
+true # ingnore failing `set -e`
```
```suggestion
true # ignore failing `set -e`
```

JBlaschke commented:
I'll rebase tomorrow... it's late right now... 

`git rebase` is right up there on my list called "stupid things to do when you're about to go to bed"

JBlaschke commented:
> Can you rebase on develop? You'll probably need to run `spack commands --update-completion` to update `spack-completion.bash` with the new `--fish` flags.

@adamjstewart I'm not sure what exactly you mean. I assume you mean to rebase "away" the merge commits keeping my fork up-to-date. Correct me if I'm wrong.

tgamblin commented:
@JBlaschke: if you are already merging, keep merging.  We will end up squashing things anyway, so it’s not worth it to rebase 132 commits. If they were factored neatly into features, we’d rebase them, but in this case we’ll almost certainly squash.

adamjstewart commented:
@JBlaschke I think all you need to do is run:
```console
$ spack commands --update-completion
```
and then merge the changes.

JBlaschke commented:
@adamjstewart done!

adamjstewart commented:
I don't think your `setup-env-test.fish` script is actually being run. That explains the low coverage of your PR. You'll need to modify `share/spack/qa/run-unit-tests` to call it. You may also need to create a `fishcov` wrapper like our `bashcov` wrapper. @tgamblin knows more about that than I do. @tgamblin shouldn't we also call the tests with zsh and dash to get more coverage of `setup-env.sh`? You may also need to add fish to the list of packages installed on the OS via `.travis.yml`.

adamjstewart commented:
We could also use unit tests for `lib/spack/spack/environment.py`. That doesn't necessarily need to be in this PR. Currently, we only test `sh`, not `csh` or `fish`.

JBlaschke commented:
@adamjstewart yes I suspected as much. The CI environment needs to be configured to install fish (I assume that codecov gets this from `.travis.yml`, right?). I was going to set up the CI pipepline on my fork, so I can test the fish codecov.

@tgamblin am I missing anything? I've only looked at the "getting started"/"quickstart" part of codecov. This is the first time I'm playing with codecov, but it looks like it's using your travis config. Let me know if you have any advice

tgamblin commented:
@JBlaschke: I’m not aware of any fish coverage tools for fish — are you?  We use kcov for bash, but I don’t think it understands fish.  Unless there’s a tool, you probably only need to worry about coverage for the python stuff.

For installing fish, you should just be able to have it installed via apt in Travis.yml.

JBlaschke commented:
> @JBlaschke: I’m not aware of any fish coverage tools for fish — are you? We use kcov for bash, but I don’t think it understands fish. Unless there’s a tool, you probably only need to worry about coverage for the python stuff.
> 
> For installing fish, you should just be able to have it installed via apt in Travis.yml.

@tgamblin I'm also not aware of any fish coverage tools (10 mins of googling hasn't found anything yet). So I'll focus on the python stuff for now.

gonsie commented:
Hi, I'm not sure if this comment should go here, but @becker33 recommended I share this knowledge.

In Fish, the caret `^` character is special (it redirect stderr). This means that if you follow the spack documentation and put a space in the spec line (e.g., `spack spec mpileaks ^emacs@23.3`) that spack won't get the trailing emacs spec and instead redirects to that as a file. 

You can fix this change in fish version 3.0 by running: `fish --features stderr-nocaret`

Perhaps we should turn this feature on by default when running the spack shell integration?

tgamblin commented:
> You can fix this change in fish version 3.0 by running: `fish --features stderr-nocaret`

@gonsie: what if we just warned about it if we detect that it is on?  I'm hesitant to require it but we could at least tell them what to add to their configuration.

gonsie commented:
@tgamblin a warning sounds good to me.

adamjstewart commented:
Ping @JBlaschke. Could you add the logic to actually run your unit tests in `run-unit-tests`? If you're busy, let me know and I can add it.

As for the `stderr-nocaret` thing, let's save that for another PR. I would like to get this merged sooner rather than later so we can add tab completion.

adamjstewart commented on share/spack/setup-env.fish:
```diff
@@ -0,0 +1,826 @@
+#################################################################################
+# This is a wrapper around the spack command that forwards calls to 'spack use'
+# and 'spack unuse' to shell functions. This in turn allows them to be used to
+# invoke dotkit functions.
```
I think dotkit has been removed now.

JBlaschke replied:
Done! I've fixed the comments to match the `setup-env.sh`.

adamjstewart commented on share/spack/setup-env.fish:
```diff
@@ -0,0 +1,826 @@
+end
+
+
+function contains_help_flags -d "checks for help (-h/--help) flags"
```
I think these flag checking functions will have to be updated like they are in #15429

JBlaschke replied:
The way I understand https://github.com/spack/spack/pull/15429, we want to make sure that `contains_help_flags` only matches `-h` and `--help` and __not__ `asdf-h`. I added this messy-looking regex `([^\S]*-h|^[\s]*-h[\s]*\$)` to take care of that (of course I document it in the preceding comment)

adamjstewart commented on share/spack/setup-env.fish:
```diff
@@ -0,0 +1,826 @@
+            return 0
+        end
+
+        # TODO: should this crash (we're clearly using fish, not bash, here)?
```
I think it shouldn't crash. For example, someone on bash may want to test the csh and fish output.

You also need to update `setup-env.sh` and `csh/spack.csh` to handle the new `--fish` flags, correct?

JBlaschke replied:
Got it! I removed the todos

adamjstewart commented:
@JBlaschke do you have time to work on wrapping up this PR? If not, I'm happy to help, but I'll probably need to rebase on develop and push commits to your repo. Just wanted to ask permission first since this PR is based on your `develop` branch.

JBlaschke commented:
> @JBlaschke do you have time to work on wrapping up this PR? If not, I'm happy to help, but I'll probably need to rebase on develop and push commits to your repo. Just wanted to ask permission first since this PR is based on your `develop` branch.

Hi @adamjstewart, so sorry to have dropped off the radar for a couple of weeks -- totally swamped at work! I'm just working my way down the stack of comments/change requests and am incorporating these in my fork now. Feel free to rebase on develop and push to my repo (do I need to give you access rights?) -- every time I tried that, it wen pear shaped (@tgamblin suggests squashing all my commits, cf his comment on Jan 29).

Edit: I've just done a `git rebase -X ours upstream/develop` and would need to force-push the changes to my repo. Should I go ahead and do that? -- this is the first time that I've used this workflow, and so I want to make sure that I don't break lots of things on your end.

JBlaschke commented:
> In Fish, the caret `^` character is special (it redirect stderr). This means that if you follow the spack documentation and put a space in the spec line (e.g., `spack spec mpileaks ^emacs@23.3`) that spack won't get the trailing emacs spec and instead redirects to that as a file.

@gonsie Thanks for letting me know.

> Perhaps we should turn this feature on by default when running the spack shell integration?
Unfortunately there is only so much we can do about this, because:

> Features will only be set on startup, so this variable will only take effect if it is universal or exported.
-- cf: https://fishshell.com/docs/current/index.html#featureflags

So if the user hasn't activated this feature by default, then they will need to escape the caret. Also turning this setting on during won't do much until a new shell is started. I think fish doesn't actually run command substitution as sub-shell:
```
$ set -U fish_features stderr-nocaret
$ status features
stderr-nocaret	off	3.0	^ no longer redirects stderr
qmark-noglob	off	3.0	? no longer globs
regex-easyesc	off	3.1	string replace -r needs fewer \'s
$ echo (status features)
stderr-nocaret	off	3.0	^ no longer redirects stderr
qmark-noglob	off	3.0	? no longer globs 
regex-easyesc	off	3.1	string replace -r needs fewer \'s
$ eval status features
stderr-nocaret	off	3.0	^ no longer redirects stderr
qmark-noglob	off	3.0	? no longer globs
regex-easyesc	off	3.1	string replace -r needs fewer \'s
```
You can see that stderr-nocaret remains "off". I can only find one solution: wrapping all of the spack commands in `fish -c <command>`:
```
$ fish -c "status features"
stderr-nocaret	on	3.0	^ no longer redirects stderr
qmark-noglob	off	3.0	? no longer globs
regex-easyesc	off	3.1	string replace -r needs fewer \'s
```
But this isn't really a solution (environment modules won't affect the "outer" shell). The only solution that I can think of is documenting this problem.

adamjstewart commented:
> Hi @adamjstewart, so sorry to have dropped off the radar for a couple of weeks -- totally swamped at work!

No worries, completely understand!

> I'm just working my way down the stack of comments/change requests and am incorporating these in my fork now. Feel free to rebase on develop and push to my repo

If you're still actively working on this I'll let you handle things. If you need any help from me let me know, otherwise let me know when you're done addressing reviews and I'll give it another review.

> Edit: I've just done a git rebase -X ours upstream/develop and would need to force-push the changes to my repo. Should I go ahead and do that?

Yes, that should be fine.

pat-s commented:
I fear this PR might get lost if it won't be merged in the near future.

JBlaschke commented:
@tgamblin @gonsie I've added a warning whenever `status feature-check stderr-nocaret` fails:
```
WARNING: you have not enabled the 'stderr-nocaret' feature.
This means that you have to escape the caret (^) character when defining specs.
Consider enabling stderr-nocaret: https://fishshell.com/docs/current/index.html#featureflags
```
Feel free to adjust the language to bring it in line with other warnings in spack.

@adamjstewart I've had trouble rebase'ing without other files (which I know I didn't touch) changing. Do you have any advice? It sounded like you have a strategy. I've given you access to my fork, in case you need to push.

adamjstewart commented:
You don’t really need to rebase, it’s fine to keep things the way they are. And don’t worry about squashing things to a single commit, I can do that automatically when I merge.

JBlaschke commented:
@adamjstewart yea I think I'll keep my grubby hands off rebase for the time being -- I will have to learn this eventually, but perhaps a simpler setting is the better learning environment.

The only thing left on my list for the PR is to add the fish unit tests. Or am I missing anything?

becker33 commented:
@JBlaschke I think that the `spack load` and `spack unload` commands need to be handled similarly to the `spack env` command to support changes to the user environment (since those commands no longer rely on `modules` under the hood.

Do you have time to update this PR soon? We're pushing towards a new major release (v0.15.0) at the end of this week and I would like to include this if possible.

JBlaschke commented:
@becker33 sure -- I'll have a look at it today.

JBlaschke commented:
@becker33 @adamjstewart @ax3l All done -- I had to add the "--fish" arg to load/unload, and I updated the QA script. Everything seems to be working fine on my end.

becker33 commented:
@JBlaschke can you use the `spack commands --update-completion` command to add the new command options to the bash completion file?

JBlaschke commented:
@becker33 crap! I forgot to re-run those, thanks for reminding me. I've just pushed the updated completions.

becker33 commented:
@JBlaschke has the `setup-env-test.fish` file been setup to actually run as part of this PR, or is it merely present and available for someone else to integrate into the CI system?

It looks like this is not tied into the CI system, and I will add it to the v0.15.0 release as "experimental" support. We can work to get this into the CI system as part of a future release (at which point it will stop being "experimental").

JBlaschke commented:
@becker33 I just added `setup-env-test.fish` to `run-shell-tests`, I also added `fish` to travis' at packages -- I didn't have the bandwidth to figure out spack's CI system, or set up my own travis instance, so I assume that that's enough. Let me know what other changes I need to make.

ax3l commented:
@JBlaschke congratulations and thank you, this is great!

cc @jkelling IIRC, you asked me at some point about Spack fish shell integration - do you like to test this? :)

jkelling commented:
@ax3l I thin a not that person, I am not using fish.

becker33 commented:
@JBlaschke I think that should be sufficient, but if not I'll take it from here to get that working.

JBlaschke commented:
thanks @becker33

JBlaschke commented:
> @JBlaschke congratulations and thank you, this is great!

I think it's time to see more fish in HPC @ax3l

becker33 commented:
I absolutely cannot get the fish tests to pass in Travis. They pass locally, but not in the CI (despite the fact they report as passing, that's a known bug).

JBlaschke commented:
@becker33 maybe you can show me how they are run in travis -- and what the error is, I've only ever run them locally

becker33 commented:
@JBlaschke if you click "details" button for the on the "Travis CI - Pull Request" status check below, and then click on the "shell" tests, you can see the output.

I believe they are just run by running `$SPACK_ROOT/share/spack/qa/run-shell-tests` in the test environment, I'm assuming that I haven't quite successfully mocked the test environment locally for testing and that's why I can't reproduce the error.

becker33 commented:
@JBlaschke I think I just figured it out: The script was adding the spack executable to `fish_user_paths` but not to PATH. And my rc files was adding Spack to the path, so it didn't cause a problem locally.

JBlaschke commented:
> @JBlaschke I think I just figured it out: The script was adding the spack executable to `fish_user_paths` but not to PATH. And my rc files was adding Spack to the path, so it didn't cause a problem locally.

@becker33  I find that confusing -- `fish_user_paths` is prepended to `PATH` in fish, so you should always be able to see the contents of `fish_user_paths` in the `PATH`

JBlaschke commented:
> > @JBlaschke I think I just figured it out: The script was adding the spack executable to `fish_user_paths` but not to PATH. And my rc files was adding Spack to the path, so it didn't cause a problem locally.
> 
> @becker33 I find that confusing -- `fish_user_paths` is prepended to `PATH` in fish, so you should always be able to see the contents of `fish_user_paths` in the `PATH`

But I am not sold to using one or the other -- my thinking has always been that `fish_user_paths` is more fish-onic than `PATH`, but if this works for you (and doesn't break anything else -- I can't find a reason why it should fail based on some quick googling) go for it.

Caveat: you're following this already: paths in `fish_user_paths` are always pre-prended to `PATH` => you can't add to `fish_user_paths` and then remove from `PATH`

becker33 commented:
@JBlaschke I find it somewhat confusing as well. If I didn't believe this was a better solution, I would look into it more.

I do think this is a better solution though, because fish_user_paths is permanent, but defining functions isn't. Which means users would need to re-run the setup script to get the function definitions, and then they would continue adding to the fish_user_paths each time they run the setup script.

adamjstewart commented:
@JBlaschke any interest in adding tab completion for fish? If so I can show you what needs to happen on the Spack-side. I don't know fish, but basically the hard part is porting `share/spack/bash/spack-completion.in` to fish. The rest of `share/spack/spack-completion.bash` is auto-generated by Spack.

adamjstewart commented:
I took a look at fish tab completion and it looks really simple and intuitive. See https://blog.gobuffalo.io/tab-completion-script-for-buffalo-in-fish-shell-c8bad3705d32 for an example. There's a lot more examples in https://github.com/fish-shell/fish-shell/tree/master/share/completions.

This will require quite a bit of modifications to Spack's internals so we can access the help messages, but it should be pretty doable. If no one else gets around to it, I'll try to take a stab at it by the end of 2020.

JBlaschke commented:
Hi @adamjstewart -- I didn't have much bandwidth these last few weeks (getting things ready for a beamtime, and wrapping things up for the AMReX folks). I can take a look at tab completion over the next couple of days (tbh -- I've never implemented them before).

Since I'm still much of a Spack novice (I really only use it to build a module here and there) how about we work together on a PR? Assuming you have time.

adamjstewart commented:
> This will require quite a bit of modifications to Spack's internals so we can access the help messages, but it should be pretty doable.

I forgot that we already do this for our rST writer, so no modifications should be necessary. It should be pretty easy to reproduce something like https://blog.gobuffalo.io/tab-completion-script-for-buffalo-in-fish-shell-c8bad3705d32 from Spack's side.

@JBlaschke if you can create a basic tab completion script for something like `spack`, `spack compiler`, and `spack compiler info` then I can generate the same kind of logic for every command automatically.
